imap_processing/hi/l1a/science_direct_event.py

```diff
@@ -75,42 +74,28 @@ def parse_direct_events(de_data: bytes) -> dict[str, list]:
     Dict[str, list]
         Parsed event data.
     """
-    de_dict = defaultdict(list)
-    binary_str_val = convert_to_binary_string(de_data)
-    for event_data in break_into_bits_size(binary_str_val):
-        # parse direct event
-        de_dict["trigger_id"].append(int(event_data[:2], 2))
-        de_dict["de_tag"].append(int(event_data[2:18], 2))
-        de_dict["tof_1"].append(int(event_data[18:28], 2))
-        de_dict["tof_2"].append(int(event_data[28:38], 2))
-        de_dict["tof_3"].append(int(event_data[38:48], 2))
+    # Each DE consists of 6-bytes. Considering the data as 3 2-byte words,
+    # each word contains the following:
+    # word_0: 2-bits of Trigger ID, upper 14-bits of de_tag
+    # word_1: lower 2-bits of de_tag, 10-bits tof_1, upper 4-bits of tof_2
+    # word_3: lower 6-bits of tof_2, 10-bits of tof_3
+    # Interpret binary blob as uint16 array and reshape to (3, n)
+    data_uint16 = np.reshape(
+        np.frombuffer(de_data, dtype=">u2"), (3, -1), order="F"
+    ).astype(np.uint16)
+
+    de_dict = dict()
+    de_dict["trigger_id"] = (data_uint16[0] >> (6 + 8)).astype(np.uint8)
```

Comment:
in above line, you were ordering data by column. Does this mean that, data_uint16 would look something like this?
```
de_data = 'word_0word_1word_2word_3word_4word5'
data_uint16 = [
    [word_0, word_3],
    [word_1, word_4],
    [word_2, word_5],
]
```
Now by getting first row, you are able to unpack trigger id and so on? Each row containing some information about DE. This is me talking out my understanding out loud.

Reply:
Yes, you have it!

Reply:
Aren't there only 3, 2-byte words from the 6-bytes of a direct event? Or do they come in pairs of 2 DEs? This might explain my previous confusion, if words 2,4,5 represent another 6-byte DE.

Reply:
Looking closer, I'm wondering if my confusion came from the use of the word... "word" 😁 , being used to describe individual bytes, while the comment describes:
>"Each DE consists of 6-bytes. Considering the data as 3 2-byte words,"

Is the following correct?:

For each DE, we only read 3x 2-byte words as values into the `data_uint16` array. 
```Python
# hex buffer data of 6 bytes:
de_data = b"\x00\x00\x00\x01\x00\x02"

# read these as 3x, 2-byte words into array
data_uint16 => [
    0, # \x00\x00 => "0b00000000"
    1, # \x00\x01 => "0b00000001"
    258, # \x01\x02 = 16**2 +2  => "0b10000010"
]
```

Then we get whatever bits we need from these 3 "words".

Reply:
It seems like everyone is saying almost the same thing, but let me try to clarify. I can never remember if numpy does rows or columns first, so I will avoid those terms and instead use first and second axis. I will see if I can clean up the comments in the code too.

The `de_data` is a binary blob with Nx6 bytes of data where N = number of direct events encoded into the binary blob. Interpreting the data as big-endian uint16 data and reshaping into a (3, -1) ndarray results in an array with shape (3, N). Indexing the first axis of that array (e.g. `data_uint16[i]`) gives the `i`th 2-bytes of data for each of the N direct events. In this way, indexing the first axis and applying bitwise operators to the resulting N-element uint16 array allows extraction of the encoded DE data without looping.

Does that help any?

Reply:
Yes! The `word_3` comment we talked about, combined with Tenzin using a 2-DE example threw me off, but this all makes sense to me now.